### PR TITLE
added decorators for Table storage.

### PR DESCRIPTION
**Pull request by pdthummar**

### CODE REVIEW

File: azure/functions/decorators/function_app.py [hunk 1]
```diff
@@ -1373,3 +1374,105 @@ def decorator():
...
+        an Azure Storage or Cosmos DB account
+        All optional fields will be given default value by function host when
+        they are parsed by function host.
+
+        Ref:
+        https://docs.microsoft.com/en-us/azure/azure-function/functions-bindings-storage-table
+
+        :param arg_name: The name of the variable that represents
+        the table or entity in function code.
+        :param connection: 	The name of an app setting or setting collection
```
Comment (YunchuWang) [inline azure/functions/decorators/function_app.py:1401]: minor: extra whitespaces after colon

File: azure/functions/decorators/function_app.py [hunk 2]
```diff
@@ -1373,3 +1374,105 @@ def decorator():
...
+        :class:`FunctionBuilder` object
+        for building :class:`Function` object used in worker function
+        indexing model. This is equivalent to defining TableInput
+        in the function.json which enables function to read a table in
+        an Azure Storage or Cosmos DB account
+        All optional fields will be given default value by function host when
+        they are parsed by function host.
+
+        Ref:
+        https://docs.microsoft.com/en-us/azure/azure-function/functions-bindings-storage-table
```
Comment (YunchuWang) [inline azure/functions/decorators/function_app.py:1397]: opening link return 404, also en-us can be removed, language will be determined by dns server
image

File: azure/functions/decorators/function_app.py [hunk 3]
```diff
@@ -1373,3 +1374,105 @@ def decorator():
...
 
         return wrap
+
+    def read_table(self,
+                   arg_name: str,
+                   connection: str,
+                   table_name: str,
+                   row_key: Optional[str] = None,
+                   partition_key: Optional[str] = None,
+                   data_type: Optional[
```
Comment (YunchuWang) [inline azure/functions/decorators/function_app.py:1384]: why take and filter are not defined as params here?
image

File: azure/functions/decorators/function_app.py [hunk 4]
```diff
@@ -1373,3 +1374,105 @@ def decorator():
...
+
+        Ref:
+        https://docs.microsoft.com/en-us/azure/azure-function/functions-bindings-storage-table
+
+        :param arg_name: The name of the variable that represents
+        the table or entity in function code.
+        :param connection: 	The name of an app setting or setting collection
+        that specifies how to connect to the table service.
+        :param table_name: The Name of the table
+        :param row_key:  The row key of the table entity to read.
```
Comment (YunchuWang) [inline azure/functions/decorators/function_app.py:1404]: minor: extra whitespaces after colon

File: azure/functions/decorators/table.py [hunk 5]
```diff
@@ -0,0 +1,47 @@
...
+    def get_binding_name() -> str:
+        return TABLE
+
+    def __init__(self,
+                 name: str,
+                 connection: str,
+                 table_name: str,
+                 row_key: Optional[str] = None,
+                 partition_key: Optional[str] = None,
+                 data_type: Optional[DataType] = None):
```
Comment (YunchuWang) [inline azure/functions/decorators/table.py:22]: ditto,  take and filter

File: azure/functions/decorators/function_app.py [hunk 6]
```diff
@@ -1373,3 +1374,105 @@ def decorator():
...
+        to a table in an Azure Storage
+        All optional fields will be given default value by function host when
+        they are parsed by function host.
+
+        Ref:
+        https://docs.microsoft.com/en-us/azure/azure-function/functions-bindings-storage-table
+
+        :param arg_name: The name of the variable that represents
+        the table or entity in function code.
+        :param connection: 	The name of an app setting or setting collection
```
Comment (YunchuWang) [inline azure/functions/decorators/function_app.py:1452]: minor: extra whitespaces after colon

File: azure/functions/decorators/function_app.py [hunk 7]
```diff
@@ -1373,3 +1374,105 @@ def decorator():
...
+        :class:`FunctionBuilder` object
+        for building :class:`Function` object used in worker function
+        indexing model. This is equivalent to defining TableOutput
+        in the function.json which enables function to write entities
+        to a table in an Azure Storage
+        All optional fields will be given default value by function host when
+        they are parsed by function host.
+
+        Ref:
+        https://docs.microsoft.com/en-us/azure/azure-function/functions-bindings-storage-table
```
Comment (YunchuWang) [inline azure/functions/decorators/function_app.py:1448]: ditto

File: tests/decorators/test_table.py [hunk 8]
```diff
@@ -0,0 +1,48 @@
...
+class TestTable(unittest.TestCase):
+    def test_table_input_valid_creation(self):
+        table_input = TableInput(name="in",
+                                 table_name="dummy_table_name",
+                                 connection="dummy_in_conn",
+                                 row_key="dummy_key",
+                                 partition_key="dummy_partition_key",
+                                 data_type=DataType.UNDEFINED)
+
+        self.assertEqual(table_input.get_binding_name(), "table")
```
Comment (YunchuWang) [inline tests/decorators/test_table.py:18]: use constant for "table"

File: tests/decorators/test_table.py [hunk 9]
```diff
@@ -0,0 +1,48 @@
...
+
+    def test_table_output_valid_creation(self):
+        table_output = TableOutput(name="out",
+                                   table_name="dummy_table_name",
+                                   row_key="dummy_key",
+                                   partition_key="dummy_partition_key",
+                                   connection="dummy_out_conn",
+                                   data_type=DataType.UNDEFINED)
+
+        self.assertEqual(table_output.get_binding_name(), "table")
```
Comment (YunchuWang) [inline tests/decorators/test_table.py:38]: use constant for "table"

File: tests/decorators/test_decorators.py [hunk 10]
```diff
@@ -1411,3 +1411,168 @@ def dummy():
...
+                        connection="dummy_in_conn",
+                        row_key="dummy_key",
+                        partition_key="dummy_partition_key",
+                        data_type=DataType.STRING)
+        def dummy():
+            pass
+
+        func = self._get_user_function(app)
+
+        output = func.get_bindings()[0]
```
Comment (YunchuWang) [inline tests/decorators/test_decorators.py:1539]: do length check on get_bindings() before this line

File: tests/decorators/test_decorators.py [hunk 11]
```diff
@@ -1411,3 +1411,168 @@ def dummy():
...
+                         connection="dummy_out_conn",
+                         row_key="dummy_key",
+                         partition_key="dummy_partition_key",
+                         data_type=DataType.STRING)
+        def dummy():
+            pass
+
+        func = self._get_user_function(app)
+
+        output = func.get_bindings()[0]
```
Comment (YunchuWang) [inline tests/decorators/test_decorators.py:1567]: do length check on get_bindings() before this line

File: azure/functions/decorators/function_app.py (hunk 1 above)
Comment (pdthummar) [reply azure/functions/decorators/function_app.py]: updated as per recommendation.

File: azure/functions/decorators/function_app.py (hunk 2 above)
Comment (pdthummar) [reply azure/functions/decorators/function_app.py]: added aka.ms link.

File: azure/functions/decorators/function_app.py (hunk 3 above)
Comment (pdthummar) [reply azure/functions/decorators/function_app.py:1386]: updated as per recommendation.

File: azure/functions/decorators/function_app.py (hunk 4 above)
Comment (pdthummar) [reply azure/functions/decorators/function_app.py]: updated as per recommendation.

File: azure/functions/decorators/table.py (hunk 5 above)
Comment (pdthummar) [reply azure/functions/decorators/table.py:24]: updated as per recommendation.

File: azure/functions/decorators/function_app.py (hunk 6 above)
Comment (pdthummar) [reply azure/functions/decorators/function_app.py]: updated as per recommendation.

File: azure/functions/decorators/function_app.py (hunk 7 above)
Comment (pdthummar) [reply azure/functions/decorators/function_app.py]: added aka.ms link

File: tests/decorators/test_table.py (hunk 8 above)
Comment (pdthummar) [reply tests/decorators/test_table.py]: updated as per recommendation.

File: tests/decorators/test_table.py (hunk 9 above)
Comment (pdthummar) [reply tests/decorators/test_table.py]: updated as per recommendation.

File: tests/decorators/test_decorators.py (hunk 10 above)
Comment (pdthummar) [reply tests/decorators/test_decorators.py]: updated as per recommendation.

File: tests/decorators/test_decorators.py (hunk 11 above)
Comment (pdthummar) [reply tests/decorators/test_decorators.py]: updated as per recommendation.

**Review by gavin-aguiar — APPROVED** [review]

**Review by YunchuWang — APPROVED** [review]

**Pull request merged**